From c4369edb03a2ad7fe32eb7cbb0ea4c48f775478f Mon Sep 17 00:00:00 2001
From: Vignesh Lal T D <vdesikachary@maxlinear.com>
Date: Tue, 23 Jul 2024 14:05:21 +0530
Subject: [PATCH] WLANRTSYS-83388 Added support for tid2linkmap teardown in
 hostap-ng cli

Issue: The certification test case EHT-4.50.1_RUN1 with hostap-ng was failed as
tid2linkmap teardown message is not captured in sniffer.

Fix: Added support for tid2linkmap teardown in hostap-ng cli.

Unit test: EHT-4.50.1_RUN1 and EHT-4.50.1_RUN2 are passed.

FIXUP to WLANRTSYS-63318 MLO changes in iwlwav-hostap-ng: ML CSA with puncturing and BSS critical update
---
 hostapd/mxl_ctrl_iface.c  | 2 +-
 hostapd/mxl_hostapd_cli.c | 6 ++++++
 hostapd/mxl_hostapd_cli.h | 3 +++
 3 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/hostapd/mxl_ctrl_iface.c b/hostapd/mxl_ctrl_iface.c
index fd28a758b..525562aa8 100644
--- a/hostapd/mxl_ctrl_iface.c
+++ b/hostapd/mxl_ctrl_iface.c
@@ -2866,7 +2866,7 @@ static int mxl_hostapd_ctrl_iface_send_t2lm_teardown_frame(struct hostapd_data *
 {
 	u8 addr[ETH_ALEN];
 
-	hapd = mxl_get_bss_index(&cmd, hapd, true, true);
+	hapd = mxl_get_bss_index(&cmd, hapd, true, false);
 	if (NULL == hapd) {
 		wpa_printf(MSG_ERROR, "CTRL: SEND_T2LM_TEARDOWN_FRAME - there is no iface with the given name");
 		return -1;
diff --git a/hostapd/mxl_hostapd_cli.c b/hostapd/mxl_hostapd_cli.c
index 052a44bcb..f4b2a996c 100644
--- a/hostapd/mxl_hostapd_cli.c
+++ b/hostapd/mxl_hostapd_cli.c
@@ -437,4 +437,10 @@ int mxl_hostapd_cli_cmd_ml_bss_critical_update(struct wpa_ctrl *ctrl, int argc,
 {
 	return mxl_wpa_ctrl_command(ctrl, "ML_BSS_CRITICAL_UPDATE");
 }
+
+int mxl_hostapd_cli_cmd_send_t2lm_teardown_frame(struct wpa_ctrl *ctrl, int argc,
+                                                    char *argv[])
+{
+	return mxl_hostapd_cli_cmd(ctrl, "SEND_T2LM_TEARDOWN_FRAME", 2, argc, argv);
+}
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
diff --git a/hostapd/mxl_hostapd_cli.h b/hostapd/mxl_hostapd_cli.h
index 7c33dc553..6fed47846 100644
--- a/hostapd/mxl_hostapd_cli.h
+++ b/hostapd/mxl_hostapd_cli.h
@@ -108,10 +108,13 @@ int mxl_hostapd_cli_cmd_send_epcs_req_frame(struct wpa_ctrl *ctrl, int argc, cha
 #endif /* CONFIG_EPCS_MXL */
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 int mxl_hostapd_cli_cmd_ml_bss_critical_update(struct wpa_ctrl *ctrl, int argc, char *argv[]);
+int mxl_hostapd_cli_cmd_send_t2lm_teardown_frame(struct wpa_ctrl *ctrl, int argc, char *argv[]);
 
 #define MXL_CLI_COMMANDS_MLO \
 	{ "ml_bss_critical_update", mxl_hostapd_cli_cmd_ml_bss_critical_update, NULL, \
 	  " = ml_bss_critical_update" }, \
+	{ "send_t2lm_teardown_frame", mxl_hostapd_cli_cmd_send_t2lm_teardown_frame, NULL, \
+	  "<bss> <mac_addr>" }, \
 
 #else
 #define MXL_CLI_COMMANDS_MLO
-- 
2.43.0

